[OTEP] Telemetry Policy (#4738)
open-telemetry/opentelemetry-specification: oteps/4738-telemetry-policy.md

in the configuration itself.

Changes become risky because the configuration encodes institutional knowledge
that is not legible to new team members or external tools.

### Configurations require global reasoning

To safely change one part, you need to understand the whole. Data flows through
a DAG of components—what shape is it at this point? What came before? What
breaks if you modify this? The cognitive load grows with the config size until
changes become risky and reviews become superficial.

When using OpAMP with an OpenTelemetry Collector, the controlling server needs
to understand the configuration layout of that specific collector. If a user
asks the server to "filter out all attributes starting with `x.`", the server
must understand and parse the collector configuration. If the same server also
manages an OpenTelemetry SDK, it needs a second implementation of the attribute
filtering feature—one for the SDK, one for the Collector. Each component has its
own configuration format and semantics.

### Configurations don't scale

Dropping a hundred noisy log patterns is feasible. A thousand patterns degrades

> performance. Ten thousand is impractical. The sequential processing model was

Reviewer: a good example of this is schema transformation in collector. This is an example of OTTL config that would transform telemetry into semconv 1.40.0 from any prev version (1.5k LoC of generated OTTL) https://github.com/open-telemetry/opentelemetry-weaver-examples/pull/36

assuming perf with policies can accommodate this, we would be able to generate them in semconv along with the release and publish next to conventions

One thing to learn from this is that a lot of dropping / renames / redactions are signal-agnostic and then need to be repeated for each target. It's not a showstopper (assuming that policies are mostly generated), but is annoying. It could potentially be solved with 'attribute'-level targets at some point in the future.

Reviewer 2: yep, this is absolutely the use case here. I think having attribute level policies will be a good extension as well!

Reviewer 2: we could also run the policies over telemetry and then verify with weaver that the transformations were successful.

component that explicitly requires dynamic behavior, there is no guarantee that
remote configuration changes take effect without a full restart.

The OpenTelemetry Collector allows custom configuration file formats. A
controlling server cannot operate with an arbitrary Collector distribution
without understanding all possible configuration formats it may encounter.

### A different model

These goals can be achieved without changing the direction of OpAMP or
file-based configuration. The solution is to separate "configuration" from
"policy".

Policies are independent rules. Each policy is atomic, self-contained, and
understandable in isolation. The execution model supports tens of thousands of
policies without degradation. A policy works the same way whether it runs in an
SDK, a Collector, or any other component that implements the specification.

## Explanation

We define a new concept called a `Telemetry Policy`. A Policy is an intent-based
specification from a user of OpenTelemetry.

- **Typed**: A policy self-identifies its "type" through its target signal. In
  the proto schema, this is enforced by the `oneof target` field — a policy

>   targets exactly one signal (e.g. log, metric, trace). Policies of different

Reviewer 3: I'm biased 😅 
```suggestion
  targets exactly one signal (e.g. log, metric, profile, trace). Policies of different
```

>   targets exactly one signal (e.g. log, metric, profile, trace). Policies of different

Reviewer 4: Why the limitation to a single signal type? Seems like it would be nice, for example, to have a policy which  redact sensitive information for all signal types.

Reviewer 2: This is mostly so that we can be deliberate about what data we are effecting. I can imagine a case where we want to apply a filter to a log for a service and attribute combo but not a trace because we want to rely on the trace over the log. It also allows the proto to constrain the grammar for the possible matching fields (i.e. if we care about affecting logs, we only have access to log matchers).

>   targets exactly one signal (e.g. log, metric, profile, trace). Policies of different

Reviewer: I can imagine policies being quite useful for  other things beyond signals, e.g. context propagators: drop / add / modify specific property in tracestate or baggage, drop context all together.

I don't see anything in the OTEP that would not allow it, just checking if extensibility like this aligns with your vision.

Reviewer 2: yep definitely on my mind, and something i kind of do currently with trace policies. Certainly baggage, resource, etc. will be something we will do.

  types cannot be merged, but policies of the same type MUST be merged together.
- **Clearly specified behavior**: A policy type enforces a specific behavior for
  a clear use case, e.g. trace sampling, metric aggregation, attribute
  filtering.
- **Implementation Agnostic**: I can use the exact same policy in the collector
  or an SDK or any other component supporting OpenTelemetry's ecosystem.

> - **Standalone**: I don't need to understand how a pipeline is configured to

Reviewer 4: You don't even need to know which components are present in a pipeline, right?  E.g. could be an SDK, or could be OBI.

Reviewer 2: yep, accurate. This point combines with the one above as well. Should I update the language here accordingly?

Reviewer 4: up to you. Just making sure I understood the intent.

  define policy.
- **Dynamic**: We expect policies to be defined and driven outside the lifecycle
  of a single collector or SDK. This means the SDK behavior needs the ability to
  change post-instantiation.
- **Idempotent**: I can give a policy to multiple components in a
  telemetry-plane safely. E.g. if both an SDK and collector obtain an
  attribute-filter policy, it would only occur once.

Every policy is defined with the following:

- A `type` denoting the use case for the policy
- A schema denoting what a valid definition of the policy entails, describing
  how servers should present the policy to customers.
- A specification denoting behavior the policy enforces
  - A specification makes clear the protobuf structure
  - The behavior that is expected for an implementation

>   - A set of examples and testcases to verify the behavior

Reviewer 5: nit:

```suggestion
  - A set of examples and test cases to verify the behavior
```

Policies MUST NOT:

- Specify configuration relating to the underlying policy applier
  implementation.

>   - A policy cannot know where the policy is going to be run.

Reviewer 5: Could you elaborate a bit more on "where the policy is going to be run"?
For example, I could imagine a policy to say: "the user email address must be removed before the telemetry leaves the customer owned machine/cluster".

Reviewer 2: This is about ensuring that a policy doesn't have configuration targeting where it will run. A user will want that exact policy, but it's up to them to deploy the policy to achieve that. I.e. the policy would say "remove the user email address from all telemetry" and they would run that in their telemetry infrastructure to ensure the guarantee that it runs prior to leaving customer owned machines/clusters. The intent of this rule is to not mix the "control plane" and the "data plane". The control plane determines where the policies run, while the data plane's purpose is just to run the policy. Policies may have labels / resource attributes to help inform the control plane on where a policy should be delegated, but that's not a requirement.

You can see in the architectural section, we will have multiple policy providers. This is the link from the data plane to the control plane and is what would allow this type of delegation to work.

Let me know if you have more questions about this, happy to elaborate further.

Reviewer 4: Maybe clarify this as "cannot know what will enforce the policy"?

Reviewer 2: Ah yes, i like that language thank you.

- Specify its transport methodology.
- Interfere with telemetry upon failure.

>   - Policies MUST be fail-open.

Reviewer 4: This seems like it would be problematic for redaction use-cases. How can I guarantee things like "EU user data doesn't leave this region", "traces don't contain PII", etc. when a bug in the policy might mean telemetry gets through"? Should be policy-dependent?

Reviewer 2: yes and no. This is more about not erroring when a field isn't available in the matching syntax. So if a user wants to redact PII for something where the attribute value is set to `cvv` but the naming convention is `user.cvv`, we shouldn't fail and cause the data flow to stop there. Similarly, these policies are ultimately compiled meaning that we should really be catching any bugs or possible failures far earlier in the process. Similarly, this guarantee of fail open behavior ensures that one policy cannot effect another policy's application. If we were to have policy dependent behavior here where one policy for EU user data staying in the region which says "fail-open" and another for PII which has "fail on missing field", if the PII one fails and we drop the data or error (stopping the data flow) then the policy for the EU user data is never applied and all that data is dropped.

Reviewer 4: > these policies are ultimately compiled meaning that we should really be catching any bugs or possible failures far earlier in the process

Can you share more about how you expect this to work? Is it like K8s where you expect your policy store to validate the policies when they are applied, so that there aren't errors when they are run by the collector/sdk/etc.?

Reviewer 2: When an implementation gets a list of policies from a provider and goes to compile them for use in the hot path it will accumulate any errors. These errors will be reported back to the provider in the next message. The behavior is to only accept policies that compile without error. It's expected that the provider _should_ do some work in advance but the implementations won't crash if they're given something invalid.

    File -->|policies| PA
    HTTP -->|policies| PA
    OpAMP -->|policies| PA
    Custom -->|policies| PA

    %% Providers supply policies to implementation
    File -->|supplies policies| PI
    HTTP -->|supplies policies| PI
    OpAMP -->|supplies policies| PI
    Custom -->|supplies policies| PI
    PA -->|supplies policies| PI

    %% Policies relationship
    PP -->|provides| policies
    PI -->|runs| policies

    %% Optional type info
    PP -.->|"may supply supported policy types (optional)"| PI
```

</details>

### Example Ecosystem Implementations

The following observations and recommendations describe how the community may
integrate with this specification.

#### OpenTelemetry SDKs

An SDK's declarative configuration may be extended to support a list of policy
providers. An SDK with no policy providers configured behaves the same as
today—policies are fail-open. The simplest policy provider is the file provider.
The SDK reads this file at startup and optionally watches for changes.

Policy providers push policies into the SDK, allowing the SDK to become a policy
implementation. An SDK may receive updates at any time, so it must support

> reloading in its extension points. Sample SDK extension points:

Reviewer: I would expect Policy sampler processors, reader, etc to be defined in the spec including per-method behavior and implementation guidance (e.g. for traces there are 3 callbacks and certain things can only happen in onEnding phase, similar logging processors have Enabled method and record-level log filtering should happen there rather than after log record is created).

I think it's ok not to go into these details in the OTEP as long as we considered implications on the design and know it's feasible to implement it.

Reviewer 2: yes we will have guidance here and will include specific implementation behaviors. We will also be shipping a suite of conformance tests to ensure that, if treating each implementation as a blackbox, those implementations all perform the same way given the same set of policies.

received from external providers.

The Collector may also serve as a policy aggregator through a policy extension.
The extension pulls from multiple policy providers while other policy
implementations set the Collector as their policy provider. This pattern enables
a horizontally scalable architecture where all extensions eventually report the
same policies.

#### OpAMP

This specification makes no requirements on the transport layer for policy
providers. OpAMP may serve as a policy provider through custom messages. A
policy implementation with OpAMP support may use the OpAMP connection to
transport policies. This specification makes no recommendation on the custom
message format.

#### Summary

This specification makes no requirements on these groups. It is recommended that
they adhere to a consistent experience for users to enhance portability.
Coordination with other SIGs will ensure agreement on configuration. A follow-up
specification may recommend policy provider specifics such as an HTTP/gRPC
definition, which would serve as a basis for custom implementations like OpAMP.
See `Future Possibilities` for more.

## Internal details

### Typed Schema

> Below is a sample for the schema of a policy, defined in the protobuf format. We

Reviewer: would the format live in the proto repo?

Reviewer 2: Yes, i believe that's the right place for it.

> make an effort to adhere to OpenTelemetry Semantic Conventions and previous

Reviewer: nit: curious what in the policies need to adhere to semconv?

Reviewer 2: this is more about the conventions for the design i.e. aligning with naming standards and design.

Reviewer: Semconv won't have any authority over policy properties - they only control telemetry schema, I'd remove this sentence, but it's really minor

specifications. Note: these proto definitions are subject to changes after this
OTEP is accepted.

```proto
message Policy {

>   // Unique identifier for this policy

Reviewer 3: Quick questions on this unique identifier:
If migrating a policy from one environment into another environment, what is the strategy or recommendation around this unique identifier?
And if comparing policies of two different environments, e.g. QA vs Production, is this unique ID considered and/or important?

Reviewer 2: the id should be a globally unique id. if you were to deploy the same policy to qa and production it would be the same id, however if they have different bodies they would have two different ids. I may not be understanding the question though.

Reviewer 3: > [..] if you were to deploy the same policy to qa and production it would be the same id, however if they have different bodies they would have two different ids.

Thanks - that is what I was looking for.

  bool case_insensitive = 21;
}
```

**TraceTarget**

```proto
message TraceTarget {
  // At least one matcher is required
  repeated TraceMatcher match = 1;

  // Probabilistic sampling configuration
  TraceSamplingConfig keep = 2;
}

message TraceMatcher {
  oneof field {
    TraceField trace_field = 1;
    AttributePath span_attribute = 2;
    AttributePath resource_attribute = 3;
    AttributePath scope_attribute = 4;
    SpanKind span_kind = 5;
    SpanStatusCode span_status = 6;
    string event_name = 7;
    AttributePath event_attribute = 8;
    string link_trace_id = 9;
  }

  oneof match {
    string exact = 10;
    string regex = 11;
    bool exists = 12;
    string starts_with = 13;
    string ends_with = 14;
    string contains = 15;
  }

  bool negate = 20;
  bool case_insensitive = 21;
}

message TraceSamplingConfig {
  float percentage = 1;          // 0-100
  string mode = 2;               // "hash_seed", "proportional", or "equalizing"
  int32 sampling_precision = 3;  // hex digits for threshold encoding (1-14)
  int32 hash_seed = 4;           // hash seed for deterministic sampling
  bool fail_closed = 5;          // reject items on sampling errors
}

> ```

Reviewer 3: Can we get a ProfileTarget?

Reviewer 2: yes, however, these protos are mostly for demonstration purposes and not finalized. they will be worked on and discussed separately after this otep is merged.

</details>
Throughout the schema, we take advantage of `oneof` to prevent invalid
configuration (i.e. someone specifying type: trace and then a metric-only
configuration).

> #### Policy Matchers

Reviewer 4: Did you consider using an existing matching language vs defining our own? I know lots of k8s things (e.g. OPA, admission policies) use CEL for matching for policies. Not sure if there are other things we should consider.

Reviewer 2: I did, i thought i put this section in but I guess it got lost. (similar to [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/4738#discussion_r2585726305)). The reason i'm defining my own matching here is very related to performance. By the scheme here, we're able to do some performance tricks to enable much more performant matching over something like CEL or OTTL. This allows policies to be added to the hot path without taking a hit no matter the amount of policies you run. I didn't include a lot about those details here because it's very implementation specific, and was going to include these details and more of an investigation of tradeoffs once we work on the specification rather than the OTEP.

Reviewer 2: FWIW, I would be more than happy to chat and show these performance tricks via slack if you're interested, not sure here is the right place.

Reviewer 4: We can discuss on slack

policies, each policy target configuration begins with setting a list of ANDed
matchers. The `LogMatcher` configuration below allows a user to easily target a
log or group of logs through any fields available to the log. A policy MUST
contain at least one matcher. Regular expressions MUST use RE2 syntax for
cross-implementation consistency.

```proto
message LogMatcher {
  // The field to match against. Exactly one must be set.
  oneof field {
    // Simple fields (body, severity_text, trace_id, span_id, etc.)
    LogField log_field = 1;

    // Log record attribute by key or path
    AttributePath log_attribute = 2;

    // Resource attribute by key or path
    AttributePath resource_attribute = 3;

    // Scope attribute by key or path
    AttributePath scope_attribute = 4;
  }

  // Match type. Exactly one must be set.
  oneof match {
    // Exact string match
    string exact = 10;

    // Regular expression match
    string regex = 11;

    ...
  }

  // If true, inverts the match result
  bool negate = 20;

  // If true, applies case-insensitive matching to all match types
  bool case_insensitive = 21;
}
```

> ### Policy Design

Reviewer: assuming we have a formal policy language in the spec / proto, I'd be interested to explore using them as a language for version transformations in semconv ([file_format](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/schemas/file_format_v1.1.0.md) and [schema v2 otep](https://github.com/open-telemetry/opentelemetry-specification/blob/main/oteps/4815-semantic-conventions-schema-v2.md#schema-transformations) for details).

It seems to be possible to do some of them with this design, but wanted to check if that's a feasiable direction. Specific examples that might be challenging:

- change span name, status code depending on attributes 
- change metric unit and update datapoint value accordingly based on metric name
- difficult one: merge values of two attributes into one.

Reviewer 2: We can definitely change span name and status code depending on attrs today. Changing unit and values is currently not supported because we'd need to think about how to ensure idempotency. Merging attributes is another one that would be challenging and something we'd need to consider how to achieve this in the future.

Author: @jaronoff97 Changing unit actually has easy idempotenency - you can just check if the unit is the one you need first.  Agree on value semantics unrelated to unit though.  Particularly in the "merging" attributes = aggregation, so we MAY only be able to support such policies either *in SDK* or with guaranteed architectural setup - which really lowers the value.

@lmolkova I think this is definitely something we should be prototyping in the policy rollout.

    system. We cannot prevent changes that may take down a process.
  - We cannot limit the execution overhead of configuration or fine-grained
    control over what changes would be allowed remotely.

### Summary

This specification draws from all of these approaches but prioritizes
independence, portability, and scale over flexibility. Where pipeline
configurations offer maximum expressiveness, policies offer predictability.
Where OPA provides a general-purpose language, policies provide a minimal,
purpose-built model. Where vendor solutions lock users in, policies use
OpenTelemetry's data model for portability.

## Open questions

What are some questions that you know aren't resolved yet by the OTEP? These may
be questions that could be answered through further discussion, implementation
experiments, or anything else that the future may bring.

## Prototypes

- [usetero/policy](https://github.com/usetero/policy)
  - The policy specification, defining the schema, matching behavior, merging
    semantics, and conformance requirements.
- [usetero/policy-go](https://github.com/usetero/policy-go)
  - Go implementation of the policy specification, designed for integration with

>     the OpenTelemetry Collector and other Go-based telemetry components.

Reviewer 6: Not a blocker on this OTEP entry specifically, but I feel I should mention that the usage of Hyperscan will make it incompatible with the OpenTelemetry Collector, at least with upstream's releases that endeavour to keep `CGO_ENABLED=0`. Perhaps the implementation could be opened up to allow for a Go native option for an alternate `CompiledDatabase`?

Reviewer 2: yep, i've already discussed this with some folks from Prometheus who are interested in this work. I think we'll have multiple options here for the internal database we can make configurable. It's a good performance trick, but obviously some consumers will not want to do this.

the configuraiton layout of an OpenTelemetry collector.  If a user asked the
server to "filter out all attributes starting with `x.`", the server would
need to understand/parse the OpenTelemetry collector configuration. If the
controlling sever was also managing an OpenTelemetry SDK, then it would need
a *second* implementation of the 'filter attribute" feature for the SDK vs.
the Collector. Additionally, as the OpenTelemetry collector allows custom
configuration file formats, there is no way for a "controlling server" to
operate with an OpenTelemetry Collection distribution without understanding all
possible implementations it may need to talk to.

Additionally, existing remote-control capabilities in OpenTelemetry are not
"guaranteed" to be usable due to specification language. For example, today
one can use the Jaeger Remote Sampler specified for OpenTelemetry SDKs and the
jaeger remote sampler extension in the OpenTelemetry collector to dynamically
control the sampling of spans in SDKs. However, File-based configuration does
not require dynamic reloading of configuration. This means attempting to
provide a solution like Jaeger-remote-sampler with just OpAMP + file-based

> config is impossible, today.

Author: I'm going to assume this is resolved - Our proposals will work *with* declarative config and augment it with policy.

  of a single collector or SDK. This means the SDK behavior needs the ability
  to change post-instantiation.
- **Idempotnent**: I can give a policy to multiple components in a
  telemetry-plane safely. E.g. if both an SDK and collector obtain an
  attribute-filter policy, it would only occur once.

Every policy is defined with the following:

- A `type` denoting the use case for the policy
- A json schema denoting what a valid definitin of the policy entails.
- TODO - A merge algorithm, denoting how multiple policies can be merged
  together in a component to create desired behavior.
- TODO - A specification denoting the behavior the policy enforces.
- TODO - *implicily* a policy has a target resource / signal it is aimed at.
  This will be used to route policies to destinations.

Example policy types include: 
- `trace-sampling`: define how traces are sampled
- `metric-rate`: define sampling period for metrics
- `log-filter`: define how logs are sampled/filtered
- `attribute-redaction`: define attributes which need redaction/removal.
- `metric-aggregation`: define how metrics should be aggregated (i.e. views).

> - `exemplar-sampling`: define how exemplars are sampled

Author: You can read some of my rationale at the bottom of the OTEP.

Effectively:

- I think policies *will* be used as-in in declarative config.  There would be a component in Declarative config that can be configured with a default set of policies.
- I think policies will be *highly limited* in expected behavior / security profile vs. declarative config.
  - I think SDK configuration will opt-in to allow remote-policy control, with explicit permissions per-policy
  - I do not think policies will alter pipeline setup or shape.  Policies should have well defined insertion points already defined via Config where they will be enforced.
- Policies will need to have a mechanism via OpAMP to advertise they can be accepted and handled - we can use "custom capabilities" for this.

So there are a lot of similarities, but the key difference is the *limitations*.

> - `exemplar-sampling`: define how exemplars are sampled

Author: @jack-berg - To resolve this - I'm going to propose we ensure all policies are unified with configuration's data model in some way that they can be expressed directly in configuration (e.g. the bootstrap policy) and also sent via OpAMP to various Otel components (OBI, Collector, SDK, etc.)

> - `exemplar-sampling`: define how exemplars are sampled

Reviewer 7: Feel free to resolve

OpenTelemetry.

## Explanation

We define a new concept called a `Telemetry Policy`.  A Policy is an
intent-based specification from a user of OpenTelemetry.

- **Typed**: A policy self identifies its "type". Policies of different types
  cannot be merged, but policies of the same type MUST be merged together.
- **Clearly specified behavior**: A policy type enforces a specific behavior for
  a clear use case, e.g. trace sampling, metric aggregation, attribute
  filtering.
- **Implementation Agnostic**: I can use the exact same policy in the collector
  or an SDK or any other component supporting OpenTelemetry's ecosystem.
- **Standalone**: I don't need to understand how a pipeline is configured to define
  policy.
- **Dynamic**: We expect policies to be defined and driven outside the lifecycle
  of a single collector or SDK. This means the SDK behavior needs the ability
  to change post-instantiation.
- **Idempotnent**: I can give a policy to multiple components in a
  telemetry-plane safely. E.g. if both an SDK and collector obtain an
  attribute-filter policy, it would only occur once.

Every policy is defined with the following:

- A `type` denoting the use case for the policy

> - A json schema denoting what a valid definitin of the policy entails.

Author: Deferring this to the actual design work that will happen post acceptance of this PR.

Design of Policy i think we'll take one at a time and have a lot of care and back-and-forth on each.

- TODO - A merge algorithm, denoting how multiple policies can be merged
  together in a component to create desired behavior.
- TODO - A specification denoting the behavior the policy enforces.
- TODO - *implicily* a policy has a target resource / signal it is aimed at.
  This will be used to route policies to destinations.

Example policy types include: 
- `trace-sampling`: define how traces are sampled
- `metric-rate`: define sampling period for metrics
- `log-filter`: define how logs are sampled/filtered
- `attribute-redaction`: define attributes which need redaction/removal.
- `metric-aggregation`: define how metrics should be aggregated (i.e. views).
- `exemplar-sampling`: define how exemplars are sampled

TODO - more examples?

TODO - Remaining high level pieces:

- SDK Components
  - `PolicyProvider`
    - Can "push" policies into the provider.

>     - Provides "observable" access to policies (e.g. notify on change)

Author: Good callout - something we'll consider when we create further prototypes here - and hope to get further input on the SDK design as it progresses.

  - Extension Points
    - `PolicySampler`: Pulls relevant `trace-sampling` policies from
      PolicyProvider, and uses them.
    - `PolicyLogProcessor`: Pulls Relevant `log-filter` policies from
      PolicyProvider and uses them.
    - `PolicyPeriodicMetricReader`: Pulls Relevant `metric-rate` policies
      from PolicyProvider and uses them to export metrics.
    - TODO: SDK-wide attribute processors

>     - TODO: SDK-view policies

Reviewer 8: Could this be simplified to instead use an applyPolicy method on the PolicyAgent. The policyprovider can use the the data (scope & policy type) from the detect method to only invoke apply on the correct audience

Author: It's a possible implementation.  This is meant to be a sketch - we'll have this as something we consider as we start prototyping SDK behavior beyond basic processors.

- Collector Components
  - `PolicyProcessor`
    - Pulls configured policies that can be enforced as a processor.
    - E.g. `log-filter`, `attribute-redaction`
  - TODO - others?
- OpAmp Interaction
  - Policy = custom extension

>   - Can we safely "roll back" a policy if it caused a breakage?

Author: Good to know - I'd like to actually implement/prototype to it so we can flesh out if we're missing anything or we have odd corner cases.

implementation or pipeline set-up.  For example, imagine a simple collector
configuration:

```yaml
recievers:
  otlp:
  prometheus:
    # ... config ...
processors:
  batch:
  memorylimiter:
  transform/drop_attribute:
    # config to drop an attribute
exporters:
  otlp:
pipelines:
  metrics/crtical:
    receivers: [otlp]
    processors: [batch, transform/drop_attribute]
    exporters: [otlp]
  metrics/all:
    receivers: [prometheus]
    processors: [memorylimiter]
    exporters: [otlp]
```

Here, we have two pipelines with intended purposes and tuned configurations.
One which will *not* drop metrics when memory limits are reached and another
that will. Now - if we want to drop a particular metric from being reported,
which pipeline do we modify?  Should we construct a new processor for that
purpose?  Should we always do so?

Now imagine we *also* have an SDK we're controlling with declarative config. If
we want to control metric inclusion in that SDK, we'd need to generate a
completely different looking configuration file, as follows:

```yaml
file_format: '1.0-rc.1'
# ... other config ...
meter_provider:
  readers:
    - my_custom_metric_filtering_reader:

>         my_filter_config: # defines what to filter

Reviewer 7: You want to filter metrics using a filtering reader (this component doesn't exist in the SDK spec and so would have to be custom) vs. views or meter config?

Author: @jack-berg Are you ok if we resolve this during the next stage of prototyping? I can work on an SDK prototype with you

Reviewer 7: I was just calling out the fact that there was more assumptions in the types of things that need to exist for this to work. Not blocking, so feel free to resolve.

        wrapped: 
          periodic:
            exporter:
              otlp_http:
                endpoint: ${OTEL_EXPORTER_OTLP_ENDPOINT:-http://localhost:4318}/v1/metric
```

Here, I've created a custom component in java to allow filtering which metrics are read.
However, to insert / use this component I need to have all of the following:

> - Know that this component exists in the java SDK

Author: That still won't tell me if it's safe to send configuration to an SDK or not.  I need to know, at runtime, that the version of the SDK I'm trying to control will support that config or if I'll crash a key component.

Additionally, it doesn't help me ignore the implementation detail.  E.g. what If I also want to control Prometheus client library?  We don't own their config or their specification.  However, we could build something that interacts with remote policies, similar to Jaeger-Remote-Sampler of today for traces.

> - Know that this component exists in the java SDK

Reviewer 9: Can the SDK be self-protecting? When the SDK receives a configuration that it does not fully support it can reject the configuration and send an error other information about the rejection back using OpAMP.

> - Know that this component exists in the java SDK

Author: Marking this is resolved - as I think your other overarching question of "how will this work with metrics SDK" is something I'd like to prototype and tracked in a different open question.

need to understand/parse the OpenTelemetry collector configuration. If the
controlling sever was also managing an OpenTelemetry SDK, then it would need
a *second* implementation of the 'filter attribute" feature for the SDK vs.
the Collector. Additionally, as the OpenTelemetry collector allows custom
configuration file formats, there is no way for a "controlling server" to
operate with an OpenTelemetry Collection distribution without understanding all
possible implementations it may need to talk to.

Additionally, existing remote-control capabilities in OpenTelemetry are not
"guaranteed" to be usable due to specification language. For example, today
one can use the Jaeger Remote Sampler specified for OpenTelemetry SDKs and the
jaeger remote sampler extension in the OpenTelemetry collector to dynamically
control the sampling of spans in SDKs. However, File-based configuration does
not require dynamic reloading of configuration. This means attempting to
provide a solution like Jaeger-remote-sampler with just OpAMP + file-based
config is impossible, today.

However, we believe there is a way to achieve our goals without changing

> the direction of OpAmp or File-based configuration. Instead we can break apart

Reviewer 10: ```suggestion
the direction of OpAMP or file-based configuration. Instead we can break apart
```

[JSON Merge Patch](https://datatracker.ietf.org/doc/html/rfc6902) to create policies that can be merged and
remain idempotent. Below we have the same update for a hypothetical `metric-rate` policy that can be merged following the RFCs

```json
# JSON Merge Patch
{
  "rpc.server.latency": { "sampling_rate_ms": 10000 }
}

# JSON Patch
[
  { "op": "add", 
    "path": "/rpc.server.latency", 
    "value": { 
      "sampling_rate_ms": 10000 
    }
  }
]
```

Proto based transmission protocols can rely on [`Merge`](https://pkg.go.dev/google.golang.org/protobuf/proto#Merge) or [`MergeFrom`](https://protobuf.dev/reference/cpp/api-docs/google.protobuf.message/#Message.MergeFrom.details) provided by the library

The mechanism for negotiating a protocol will depend on the specific `PolicyProvider` implementation, some options are:

* A `FileProvider` will either use a default merger from the format (like the default proto merge), or accept a parameter that specifies which merger is expected when reading the specific file format (for example, for JSON).
* A HTTP provider can use different file formats to decide which merger to use, as specified in the RFCs for JSON patch formats.

> * OpAmp providers could add a field specifying the merger as well as the data being transmitted, plus a mechanism for systems to inform each other which mergers are avaliable and how the data is expected to be merged.

Reviewer 10: ```suggestion
* OpAMP providers could add a field specifying the merger as well as the data being transmitted, plus a mechanism for systems to inform each other which mergers are avaliable and how the data is expected to be merged.
```

This is not ideal for a few reasons:

- Anyone designing a server that can control telemetry flow MUST have a deep
  understanding of all components it could control and their implementations.
- We don't have a "safe" mechanism to declare what configuration is supported
  or could be sent to a specific component (note: we can design one)
- The level of control we'd expose from our telemetry systems is *expansive*
  and possibly dangerous. 
  - We cannot limit the impact of any remote configuration on the working of a
    system. We cannot prevent changes that may take down a process.
  - We cannot limit the execution overhead of configuration or fine-grained
    control over what changes would be allowed remotely.

## Open questions

What are some questions that you know aren't resolved yet by the OTEP? These may be questions that could be answered through further discussion, implementation experiments, or anything else that the future may bring.

- Should this specification give recommendations for the server protobufs
- How should we handle policy merging?
  - (jacob) Could policies contain a priority and it's up to the providers to design around this?

> 

Reviewer 11: You raise very good points about OTTL's lack of grammar and stability guarantees. Also limiting policy types to OTTL is probably too restrictive - this was my suspicion too. Thank you for bringing these arguments.

> 

Author: Marking this as resolved

provide a solution like Jaeger-remote-sampler with just OpAMP + file-based
config is impossible, today.

However, we believe there is a way to achieve our goals without changing
the direction of OpAmp or File-based configuration. Instead we can break apart
the notion of "Configuration" from "Policy", providing a new capability in
OpenTelemetry.

## Explanation

We define a new concept called a `Telemetry Policy`.  A Policy is an
intent-based specification from a user of OpenTelemetry.

- **Typed**: A policy self identifies its "type". Policies of different types
  cannot be merged, but policies of the same type MUST be merged together.
- **Clearly specified behavior**: A policy type enforces a specific behavior for
  a clear use case, e.g. trace sampling, metric aggregation, attribute
  filtering.
- **Implementation Agnostic**: I can use the exact same policy in the collector
  or an SDK or any other component supporting OpenTelemetry's ecosystem.
- **Standalone**: I don't need to understand how a pipeline is configured to define
  policy.
- **Dynamic**: We expect policies to be defined and driven outside the lifecycle

>   of a single collector or SDK. This means the SDK behavior needs the ability

Reviewer 9: I would really love to see things are getting more dynamic in the SDKs to be able to have more possibilities for remote configuration support.

  to change post-instantiation.
- **Idempotent**: I can give a policy to multiple components in a
  telemetry-plane safely. E.g. if both an SDK and collector obtain an
  attribute-filter policy, it would only occur once.

> Every policy is defined with the following:

Author: Going to defer this to our design on policies here.  It's a great question - again I think this comes down to matchers we have for policy.

```python
def post_merge(merged):
  for metric in merged:
    # If there is a new recommended sampling rate
    if metric.sampling_rate_ms is not None and
        has(metric.sampling_rate_ms.recommended):
      if not(has(metric.sampling_rate_ms.minimum)):
        # If there was no minimum, just use the new recommended one
        metric.sampling_rate_ms.minimum = metric.sampling_rate_ms.recommended
      else:
        # Otherwise, keep the smallest one.
        metric.sampling_rate_ms.minimum = min(
            metric.sampling_rate_ms.minimum,
            metric.sampling_rate_ms.recommended)
    # Cleanup: remove the new recommendation after processing
    del(metric.sampling_rate_ms.recommended)
```

This guarantees that policies can be applied in any order and yet the end result
will be the same. It also allow for more complex policies like taking into
account timestamps and sources, provided the operations can be demonstrated to
be commutative and applied in any order. For example, we could have a policy
where a given provider value always overrides another provider if one of the
fields is the provider name and the post-merge algorithm takes this information

> into account.

Reviewer 2: yup this makes sense to me, I've been working through the same thought/issue, and would be good to mandate this.

telemetry processing. Controlling telemetry behavior at scale remains difficult.
The current model—configuration files that define processing pipelines—breaks
down in predictable ways.

### Configurations grow organically

Processing rules accumulate over time. A configuration that started as 50 lines
becomes thousands. Each line represents a problem that was solved, but the
context is lost. Line 847 exists for a reason, but that reason is not documented
in the configuration itself.

Changes become risky because the configuration encodes institutional knowledge
that is not legible to new team members or external tools.

### Configurations require global reasoning

To safely change one part, you need to understand the whole. Data flows through
a DAG of components—what shape is it at this point? What came before? What
breaks if you modify this? The cognitive load grows with the config size until
changes become risky and reviews become superficial.

When using OpAMP with an OpenTelemetry Collector, the controlling server needs
to understand the configuration layout of that specific collector. If a user
asks the server to "filter out all attributes starting with `x.`", the server

> must understand and parse the collector configuration. If the same server also

Reviewer 12: Why should the server parse configuration? The server is supposed to generate configuration and send to the Collector, not "parse" it. I am confused.

Author: "It does need to know the schema of the config" - I'm not sure how "parsing" and "understanding the schema" are different.  The UI needs to take a blob of bytes and understand it - in my head that's parsing - I think we may just be talking past each other there.

I guess what you're suggesting is I could pass down config to a collector that always overwrites what it had locally.  I agree that's an alternative, but not what we're proposing here - it also fits into my main concern that you'd be tying 1:1 your UI code with the thing you're controlling.  IF the goal for OpAMP is to be used for SDKs and Collector - and hopefully other things, like promtheus clients e.g., then we need a mechanism where I can send something that is *interpreted* by the client vs. its raw config.

I'm going to mark this resolved as I think we agree - but I see the nuance you're suggesting here.

manages an OpenTelemetry SDK, it needs a second implementation of the attribute
filtering feature—one for the SDK, one for the Collector. Each component has its
own configuration format and semantics.

### Configurations don't scale

Dropping a hundred noisy log patterns is feasible. A thousand patterns degrades
performance. Ten thousand is impractical. The sequential processing model was
not designed for this level of specificity. Organizations compromise by dropping
broad categories, losing signal with the noise.

### Remote control lacks guarantees

Existing remote-control capabilities in OpenTelemetry are not guaranteed to be
usable. The Jaeger Remote Sampler works with OpenTelemetry SDKs and the
Collector's Jaeger remote sampler extension. However, file-based configuration
does not require dynamic reloading. A solution combining OpAMP with file-based

> configuration cannot provide the same dynamic behavior.

Author: I can implement the OpAMP spec + Configuration spec without the *requirement* to support dynamic adaptation, right?

That's what this is saying.  There's nothing *required* in OpAMP to allow this to work, we need a component that *requires* dynamic behaivor.

There's nuances to this point as well, but the basic gist is - We aren't requiring dynamic loading, so we'd need to change our stance or create something that DOES require it.

## Explanation

We define a new concept called a `Telemetry Policy`. A Policy is an intent-based
specification from a user of OpenTelemetry.

- **Typed**: A policy self-identifies its "type" through its target signal. In
  the proto schema, this is enforced by the `oneof target` field — a policy
  targets exactly one signal (e.g. log, metric, trace). Policies of different
  types cannot be merged, but policies of the same type MUST be merged together.
- **Clearly specified behavior**: A policy type enforces a specific behavior for
  a clear use case, e.g. trace sampling, metric aggregation, attribute
  filtering.
- **Implementation Agnostic**: I can use the exact same policy in the collector
  or an SDK or any other component supporting OpenTelemetry's ecosystem.
- **Standalone**: I don't need to understand how a pipeline is configured to
  define policy.
- **Dynamic**: We expect policies to be defined and driven outside the lifecycle
  of a single collector or SDK. This means the SDK behavior needs the ability to
  change post-instantiation.
- **Idempotent**: I can give a policy to multiple components in a
  telemetry-plane safely. E.g. if both an SDK and collector obtain an

>   attribute-filter policy, it would only occur once.

Author: To finish this - I do think we need more metadata-tagging capabilities in OTLP - particularly around ensuring behaviors like this work.  I think initially we'll start with policies where idempotency can be guaranteed, but perhaps in the future we propose OTLP additions of batch / log metadata so we can tag operations having been performed in a way these will be preserved between components.

I consider that follow on design work - the real question of this OTEP is if we want to move forward with Policy as general thing so we can start asking/answering these fundamental questions.

Every policy is defined with the following:

- A `type` denoting the use case for the policy
- A schema denoting what a valid definition of the policy entails, describing
  how servers should present the policy to customers.
- A specification denoting behavior the policy enforces, i.e., for a given JSON
  entry, to which elements the policy applies and which behavior is expected
  from an agent or collector implementing the policy.

Policies MUST NOT:

- Specify configuration relating to the underlying policy applier
  implementation.
  - A policy cannot know where the policy is going to be run.
- Specify its transport methodology.
- Interfere with telemetry upon failure.
  - Policies MUST be fail-open.
- Contain logical waterfalls.
  - Each policy's application is distinct from one another and at this moment
    MUST NOT depend on another running. This is in keeping with the idempotency
    principle.

Example policy types include:

> - `trace-sampling`: define how traces are sampled

Author: I think we could accomplish this - it depend on how we construct the policy definition to begin with.

Again - each policy has SOME control over how "merging" works - so we could do this.  Would be worth investigating and putting together a prototype.

> - `trace-sampling`: define how traces are sampled

Reviewer 4: Maybe this would work with consistent sampling?

> - `trace-sampling`: define how traces are sampled

Reviewer 2: @dashpole as implemented it already does in each of my implementations :)

> - `trace-sampling`: define how traces are sampled

Author: Marking as resolved - given the demo.

- `metric-rate`: define sampling period for metrics
- `log-filter`: define how logs are sampled/filtered
- `attribute-redaction`: define attributes that need redaction/removal.
- `metric-aggregation`: define how metrics should be aggregated (i.e. views).
- `exemplar-sampling`: define how exemplars are sampled
- `attribute-filter`: define data that should be rejected based on attributes

## Policy Ecosystem

Policies are designed to be straightforward objects with little to no logic tied
to them. Policies are also designed to be agnostic to the transport,

> implementation, and data type. It is the goal of the ecosystem to support

Reviewer 12: What does agnostic to data type mean? Does it mean a sampler policy will work for metrics, traces and logs in some agnostic way?

Author: A few things:

- This means we don't want "policy" tied to a specific `{Signal}Provider` in an SDK. I may have *different policy types* per signal, but I don't have a completely different mental model of how they work or are distributed per-signal, and I *could* use the same policy between signal types. (e.g. as you state if we wanted a sampling policy to apply uniformly across signal types)
- This means we can add a new policy definition without breaking any existing policy definitions in the ecosystem.

Author: Marking this resolved

> policies in various ways. Policies MUST be additive and MUST NOT break existing

Author: Marking this resolved

The following observations and recommendations describe how the community may
integrate with this specification.

#### OpenTelemetry SDKs

An SDK's declarative configuration may be extended to support a list of policy
providers. An SDK with no policy providers configured behaves the same as
today—policies are fail-open. The simplest policy provider is the file provider.
The SDK reads this file at startup and optionally watches for changes.

Policy providers push policies into the SDK, allowing the SDK to become a policy
implementation. An SDK may receive updates at any time, so it must support
reloading in its extension points. Sample SDK extension points:

- `PolicySampler`: Pulls relevant `trace-sampling` policies from PolicyProvider.
- `PolicyLogProcessor`: Pulls relevant `log-filter` policies from
  PolicyProvider.
- `PolicyPeriodicMetricReader`: Pulls relevant `metric-rate` policies from
  PolicyProvider.

#### OpenTelemetry Collector

The Collector is a natural place to run policies. A policy processor may be
introduced to execute policies. The Collector should use the same declarative
configuration as the SDK for policy provider configuration. The Collector may

> introduce an inline policy provider for default policies in addition to those

Reviewer 12: What is an "inline" provider?

Author: Effectively the "default" policy is provided in the configuration file of the collector.  That's all this is saying.

Author: I'm going to mark this resolved for now - unless you have suggested wording to improve it.

a horizontally scalable architecture where all extensions eventually report the
same policies.

#### OpAMP

This specification makes no requirements on the transport layer for policy
providers. OpAMP may serve as a policy provider through custom messages. A
policy implementation with OpAMP support may use the OpAMP connection to
transport policies. This specification makes no recommendation on the custom
message format.

#### Summary

This specification makes no requirements on these groups. It is recommended that
they adhere to a consistent experience for users to enhance portability.
Coordination with other SIGs will ensure agreement on configuration. A follow-up
specification may recommend policy provider specifics such as an HTTP/gRPC
definition, which would serve as a basis for custom implementations like OpAMP.
See `Future Possibilities` for more.

## Internal details

### Typed Schema

Below is a sample for the schema of a policy, defined in the protobuf format. We
make an effort to adhere to OpenTelemetry Semantic Conventions and previous
specifications.

```proto
message Policy {
  // Unique identifier for this policy

>   string id = 1;

Author: I think we can debate id vs. name when we do concrete policy design post OTEP.  The key here is to agree on any requirements for that full design.

  // Human-readable name
  string name = 2;

  // Optional description
  string description = 3;

  // Whether this policy is enabled
  bool enabled = 4;

  // Timestamp when this policy was created (Unix epoch nanoseconds)
  fixed64 created_at_unix_nano = 5;

  // Timestamp when this policy was last modified (Unix epoch nanoseconds)
  fixed64 modified_at_unix_nano = 6;

  // Labels for metadata and routing
  repeated opentelemetry.proto.common.v1.KeyValue labels = 7;

  // Target configuration. Exactly one must be set.
  oneof target {
    LogTarget log = 10;
    MetricTarget metric = 11;

>     TraceTarget trace = 12;

Reviewer 12: Can you please show what LogTarget/MetricTarget/TraceTarget messages look like?

Author: Marking this resolved

```proto
message LogMatcher {
  // The field to match against. Exactly one must be set.
  oneof field {
    // Simple fields (body, severity_text, trace_id, span_id, etc.)
    LogField log_field = 1;

    // Log record attribute by key or path
    AttributePath log_attribute = 2;

    // Resource attribute by key or path
    AttributePath resource_attribute = 3;

    // Scope attribute by key or path
    AttributePath scope_attribute = 4;
  }

  // Match type. Exactly one must be set.
  oneof match {
    // Exact string match
    string exact = 10;

    // Regular expression match
    string regex = 11;

    ...
  }

  // If true, inverts the match result
  bool negate = 20;

  // If true, applies case-insensitive matching to all match types
  bool case_insensitive = 21;
}
```

### Policy Design

The configuration for the actions for policies will be determined after this
OTEP is accepted and is currently being developed. Each policy MUST specify its
runtime requirements. Policy actions MUST be run in designated stages. When
designing a new policy, a policy SHOULD begin with a filter to select the
targeted data.

### Runtime Requirements

#### Evaluation

> Implementations MAY evaluate policies concurrently. The independence of policies

Reviewer 12: So this rules out the "multiply by 100" example I was asking about, right?

Author: I actually had some ideas how to do this in an idempotent way - but we'll save that for further work as it's a bit future focused, and we have much more important policies to push for now.

supply a policy with the same `id`, the client must decide which one to keep.
Implementations SHOULD assign each provider a priority — for example, OPAMP (1),
HTTP (2), FILE (3), CUSTOM (user-defined) — where a lower number is higher
priority. When two policies share the same `id`, the policy from the
higher-priority provider wins and the other is dropped. Where a policy from a
lower-priority provider cannot be merged consistently with the higher-priority
version, the lower-priority policy SHOULD be dropped in its entirety.

The specific mechanism will depend on the `PolicyProvider` implementation:

- A `FileProvider` reads the full policy set from disk (YAML, JSON, or proto
  binary). Each read produces a complete snapshot; no patch semantics are
  needed.
- An HTTP or gRPC provider SHOULD implement request/response sync with
  hash-based change detection and support for client metadata (supported policy
  stages, resource attributes).
- OpAMP providers can embed the policy set in an OpAMP custom-message or
  agent-config payload, reusing OpAMP's existing change-detection mechanisms.

#### Runtime conflict resolution

> 

Reviewer 13: This section assumes perfect transport order, or at least doesn't make it clear to me what happens with out of order updates.
For example policy "export-traces: on" is sent, but get's delayed. "export-traces: off" is sent and received, the agent turns off export of traces. Then the delayed "export-traces: on" gets pushed through. I have 2 problems now
1. The order of application is clearly wrong, the intention is to have no exporting but the last policy turns it on
2. Even if I wanted to combine these, it's impossible to come up with a commutative reduction for these two operations

I would favour a naive last timestamp approach (since we've got the prioritization, it's reasonable to assume that a given policy source has a specific consistent timestamper)

  telemetry, every instance MUST produce the same output.

These properties ensure that policies can be distributed across agents and
collectors without coordination, and that the outcome is reproducible regardless
of processing order.

As a concrete example, consider how a runtime might resolve conflicting `keep`
values. A naïve approach — last write wins — violates commutativity:

```python
# Bad: result depends on processing order
def resolve_keep_naive(matching_policies):
    result = "all"
    for policy in matching_policies:
        result = policy.keep  # last one wins
    return result
```

Instead, the runtime can apply a **commutative reduction** that always converges
to the same answer. For `keep`, a natural choice is "most restrictive wins":

```python
def resolve_keep(matching_policies):
    """Resolve conflicting keep values.

    Applies a commutative 'most restrictive' merge:
      none < N/s < N/m < N% < all

    The result is independent of policy ordering.
    """
    result = Keep("all")
    for policy in matching_policies:
        if policy.keep is None:
            continue
        candidate = Keep(policy.keep)
        # Keep the more restrictive of the two

>         result = most_restrictive(result, candidate)

Reviewer 2: updated the code to make this less hand wavey

open-extension and definitions of Declarative Config to pass the whole behavior
of an SDK or Collector from an OpAMP "controlling server" down to a component
and have them dynamically reload behavior.

What this solution doesn't do is answer how to understand what config can be
sent to what component, and how to drive control / policy independent of
implementation or pipeline set-up. For example, imagine a simple collector
configuration:

```yaml
receivers:
  otlp:
  prometheus:
    # ... config ...
processors:
  batch:
  memorylimiter:
  transform/drop_attribute:
    # config to drop an attribute
exporters:
  otlp:
pipelines:
  metrics/critical:
    receivers: [otlp]
    processors: [batch, transform/drop_attribute]
    exporters: [otlp]
  metrics/all:
    receivers: [prometheus]
    processors: [memorylimiter]
    exporters: [otlp]
```

Here, we have two pipelines with intended purposes and tuned configurations. One
which will _not_ drop metrics when memory limits are reached and another that
will. Now - if we want to drop a particular metric from being reported, which
pipeline do we modify? Should we construct a new processor for that purpose?
Should we always do so?

> 

Reviewer 12: OK, I think it makes sense to me. The follow up question is what is the process through which the policies are designed, published and implemented in processors and what happens if a new policy is published after the processor is released. Is there a versioning concept, a check of supported policies in the processor, etc.

I assume we don't expect policies to be expressed in some general purpose "language" that the processor implements once and all policies released in the future just work without needing to update the processor implementation (I can indeed imagine an approach like that, but this OTEP doesn't seem to suggest that).

> 

Author: Yes - the process for how to define them, etc. is what I suspect will be follow on work to this OTEP.

This OTEP motivates + gets agreement that this is a useful concept to spend time on - follow up work would propose how to formalize them, *where* to formalize them, how to send them via OpAMP, how to interact with them in SDKs, etc.

I'd like to start decomposing that work, or further prototyping important aspects with a very clear early milestone based on work @jaronoff97 has done.

> 

Author: Marking this resolved - under the assumption / declaration in this OTEP that follow upwork includes answering these questions.
